### PR TITLE
Make it possible to configure NULL representation.

### LINKED ISSUE #193: null to text conversion in array
Hello, suppose I have an array of string where some elements are nulls. The result of select of the array is:
```
epgsql:equery(C, "SELECT $1", [[<<"s1">>, null]]).        
{error,{error,error,<<"42P18">>,indeterminate_datatype,
              <<"could not determine data type of parameter $1">>,
              [{file,<<"postgres.c">>},
               {line,<<"1336">>},
               {routine,<<"exec_parse_message">>},
               {severity,<<"ERROR">>}]}}
```
cast fixes above error
```
epgsql:equery(C, "SELECT $1::TEXT[]", [[<<"s1">>, null]]).
{ok,[{column,<<"text">>,{array,text},1009,-1,-1,1}],
    [{[<<"s1">>,<<"null">>]}]}

```
but in the result the null value is also converted to string. How to avoid above error  and preserve type of null values?

### PR DESCRIPTION
**Pull request by seriyps**
New `nulls` `connect/1` option added that can be used to configure the set of terms to be used to represent NULL in input (parameters) and output (data rows).

Also, NULLs inside arrays are fixed #193

My original intention was to fix #193, but I found that it would be easier to add configurable nulls #203 at the same time.

It might affect performance slightly (because we are using `lists:member/2` instead of function clause match in a quite hot path, but the difference should be proportional to the size of `nulls` lists and it's up to user to decide on trade-offs. If all the elements of `nulls` are atoms, `lists:member` should be quite quick.

Implementation is not super clean, because there are many places where we do NULL encoding/decoding: `record` type, `array` type and generic case. Probably can be refactored later. But all the cases are covered by tests.

Some usage examples are in tests: `epgsql_SUITE:custom_null`

### CODE REVIEW

File: src/epgsql_binary.erl [hunk 1]
```diff
@@ -213,44 +245,73 @@ decode_record1(<<Oid:?int32, Len:?int32, ValueBin:Len/binary, Rest/binary>>, Siz
...
-    {Data, {NDims, Lengths}} = encode_array(Array, 0, [], ValueEncoder),
+encode_array(Array, Oid, ArrayEncoder) ->
+    {Data, {NDims, Lengths, HasNull}} = encode_array_dims(Array, ArrayEncoder),
     Lens = [<<N:?int32, 1:?int32>> || N <- lists:reverse(Lengths)],
-    Hdr  = <<NDims:?int32, 0:?int32, Oid:?int32>>,
+    HasNullInt = case HasNull of
+                     true -> 1;
+                     false -> 0
+                 end,
+    Hdr  = <<NDims:?int32, HasNullInt:?int32, Oid:?int32>>,
```
Comment (seriyps) [inline src/epgsql_binary.erl]: before this change we [always](https://github.com/epgsql/epgsql/commit/688a55dc0c1faf6eb93f252e81478269f150b7b8) had this `HasNullInt` flag set to `0` which violates PostgreSQL server expectations, but seems server still was able to handle that.

https://github.com/postgres/postgres/blob/177be9edf4bb966400db7769d61e479aa0fe0201/src/backend/utils/adt/arrayfuncs.c#L1434

File: src/epgsql_wire.erl [hunk 2]
```diff
@@ -246,16 +246,18 @@ encode_parameters([P | T], Count, Formats, Values, Codec) ->
...
-encode_parameter({T, undefined}, Codec) ->
-    encode_parameter({T, null}, Codec);
-encode_parameter({_, null}, _Codec) ->
-    {1, <<-1:?int32>>};
-encode_parameter({{unknown_oid, _Oid}, Value}, _Codec) ->
-    {0, encode_text(Value)};
 encode_parameter({Type, Value}, Codec) ->
-    {1, epgsql_binary:encode(Type, Value, Codec)};
-encode_parameter(Value, _Codec) ->
-    {0, encode_text(Value)}.
```
Comment (seriyps) [inline src/epgsql_wire.erl]: this clause was unreachable because `Value` is always 2-tuple

File: src/epgsql_binary.erl (hunk 1 above)
Comment (seriyps) [reply src/epgsql_binary.erl]: Seems this flag is validated but then ignored by server: https://github.com/postgres/postgres/blob/177be9edf4bb966400db7769d61e479aa0fe0201/src/backend/utils/adt/arrayfuncs.c#L1304-L1308

**Comment by davidw** [comment]
You might add a brief comment in the README like

"As documented below, if this parameter is not passed in, the default is `null`"

**Comment by seriyps** [comment]
Done. Thanks!

**Pull request merged**